docs/dag-sharding-rfc.md

```diff
@@ -0,0 +1,176 @@
+
+One upside is that the interaction between the ipfs files dag and cluster-dag importing process becomes much more simple.  The original importer could run to completion uninterrupted, and then the node could build the cluster-dag and unpin the leaves from the ipfs files dag.  Althought this wouldn't directly address the issues brought up above it is possible collaborative importing would result in a simpler download process as now downloads and imports are not interleaved.  It also seems likely that, if done right, collaborative importing could be faster overall, as importing would be parallelized across nodes.
+
+
+## Arbitrary dags
+
+TODO -- I have some thoughts on these (many voiced in my looong comment in response to @hsanjuan's proposal), I need to take some time to organize and record them here
+
+### Dags too big to fit in ipfs repo
+
+### Dags with data not in the leaves
+
+### Dags already imported
+
+
+# Workplan
+
+There is a fair amount that needs to be built, updated and serviced to make all of this work.  Here I am focusing on basic sharding and allowing support for files that wouldn't fit in an ipfs repo.  Fault tolerance and arbitrary dag work should come after this is nailed down.  From current design fault tolerance will be straightforward on top of basic sharding and arbitrary dag support will probably be quite different and difficult regardless of how we implement the basics.
+
+* specialized dag building tools 
+  * cluster-dag ipld-format
+  * cluster-dag importer
+  * fundamental reimagining of go-ipfs dag building
+    * Either ability to build multiple dags at once from two importers (single importer, most likely choice)
+    * Or ability to combine together the dags of arbitrarily split leaves into the dag that would have arisen from importing all leaves at once (collaborative importers, less likely choice)
+    * Could be the basis for an independent dagger library/tool
+* metric sweeps
+  * design metrics (rpc and/or ipfs remote pin unpin, and state size) and sweeps
+  * if we want better performance then we can optimize (rpc) or build new transfer tools out of libp2p
+* improving state component to scale (@hsanjuan from your proposal open questions)
+* simple updates to state format to allow for cluster-dag support
+* probably should work on improving migration framework as significant state changes are in the works
+* adding functionality to ipfs-cluster add ipfs proxy endpoint
+* changing configs and user commands to include `--shard` on add, in general figure out UX to determine when we shard
+* potentially significant modifications to cluster peers signalling that a cid should be pinned (now they may be transferring data via rpc before hand) and so potentially significant modifications to how peers interact with state.  (@hsanjuan, this is something I haven't thought through enough to convince myself it won't be too different from now so I might be overstating the difficulty here)
+
```

Comment:
I think of a work plan little by little. For example, at first the importer does not need to do cluster-sharding, it just needs to work (like ipfs does). So:

* Need to start the `importers` project. Figure out what ipfs and ipfs-pack do to chunk data and create dags and create a library that does that. Paying attention that as data is streamed in, dags nodes are streamed out.
* Particularly important to figure out how unixFS file/folder data can be streamed (via reading from disk directly, vs. via http POST multipart upload...)
* cluster-Importer Component: figure out the interface it provides and implement it without the cluster-sharding.
* Using the importers project, the cluster-importer component should be able to split data, object/put it in local ipfs, and pin it in cluster. Figure out what API cluster provides to use the importer component. Including REST API part.
* Once that works, a second round on the importer allows it to build a cluster dag, object/put it in local ipfs and pin it in cluster (need to design cluster-dag format).
* Once that works, the state format is upgraded to include the cluster dag.
* Once that works, the importer allocates different cluster-dag-parts to different cluster-peers and ipfs-nodes.
* Once that works: scaling the state.
* Usability features etc...